### PR TITLE
BODS-5435: AVL Subscription Migration Script

### PR DESCRIPTION
**Pull request by DeannaSharma**

### CODE REVIEW

File: cli-helpers/src/commands/migrate-avl-subscriptions.ts [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+        }
+
+        const subscriptionsJson = csvToJson
+            .fieldDelimiter(",")
+            .getJsonFromCsv(`${fileName}.csv`) as unknown as BodsSubscription[];
+
+        const setOfSubscriptionUrlAndUsername = new Set();
+        const dataProducersToSubscribeTo: Subscription[] = [];
+
+        subscriptionsJson.map((subscription) => {
```
Comment (nbetts) [inline cli-helpers/src/commands/migrate-avl-subscriptions.ts:75]:
```suggestion
        subscriptionsJson.forEach((subscription) => {
```

File: cli-helpers/src/commands/migrate-avl-subscriptions.ts [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+                    id: subscription.dataset_id,
+                    publisherId: subscription.organisation_id,
+                    url: subscription.url_link,
+                    username: subscription.username,
+                    password: subscription.password,
+                });
+            }
+        });
+
+        subscriptionsJson.map((subscription) => {
```
Comment (nbetts) [inline cli-helpers/src/commands/migrate-avl-subscriptions.ts:89]:
```suggestion
        subscriptionsJson.forEach((subscription) => {
```

File: cli-helpers/src/commands/migrate-avl-subscriptions.ts [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+
+        const subscriptionsJson = csvToJson
+            .fieldDelimiter(",")
+            .getJsonFromCsv(`${fileName}.csv`) as unknown as BodsSubscription[];
+
+        const setOfSubscriptionUrlAndUsername = new Set();
+        const dataProducersToSubscribeTo: Subscription[] = [];
+
+        subscriptionsJson.map((subscription) => {
+            if (subscription.status === "live") {
```
Comment (nbetts) [inline cli-helpers/src/commands/migrate-avl-subscriptions.ts:76]: I know the source data is an export from the bods db, so there shouldn't be any duplicated live rows, but I think it is worth adding in the set check (like in the loop below) to be double sure we don't attempt to subscribe to a feed multiple times

File: cli-helpers/src/commands/migrate-avl-subscriptions.ts [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+                    publisherId: subscription.organisation_id,
+                    url: subscription.url_link,
+                    username: subscription.username,
+                    password: subscription.password,
+                });
+            }
+        });
+
+        subscriptionsJson.map((subscription) => {
+            if (
```
Comment (nbetts) [inline cli-helpers/src/commands/migrate-avl-subscriptions.ts:90]: I think it's worth adding a comment in here about _why_ we loop over the list again analysing the error status subscriptions, as it's not obvious from reading the code alone

File: cli-helpers/src/commands/migrate-avl-subscriptions.ts [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+            });
+
+            if (!subscribeEvent || !subscribeEvent.Payload) {
+                unsuccessfulSubscriptions.push(subscription);
+                return;
+            }
+
+            const returnPayload: Uint8ArrayBlobAdapter = subscribeEvent.Payload;
+
+            const returnPayloadJson = JSON.parse(returnPayload?.transformToString() ?? "{}");
```
Comment (nbetts) [inline cli-helpers/src/commands/migrate-avl-subscriptions.ts:131]:
```suggestion
            const returnPayloadJson = JSON.parse(subscribeEvent.Payload?.transformToString() ?? "{}");
```

Doing it inline like this means we can avoid installing and using the `@smithy/util-stream` package

Same for the rollback script

File: cli-helpers/src/commands/migrate-avl-subscriptions.ts (hunk 1 above)
Comment (DeannaSharma) [reply cli-helpers/src/commands/migrate-avl-subscriptions.ts:75]: replaced with for of as our linter doesn't like forEach!

**Review by nbetts — APPROVED** [review]

**Pull request merged**